src/transaction/log_recovery_analysis_corruption_checker.hpp

```diff
@@ -0,0 +1,157 @@
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#ifndef _LOG_RECOVERY_ANALYSIS_CORRUPTION_CHECKER_CPP_HPP_
+#define _LOG_RECOVERY_ANALYSIS_CORRUPTION_CHECKER_CPP_HPP_
+
+#include "error_manager.h"
+#include "log_common_impl.h"
+#include "log_impl.h"
+#include "log_lsa.hpp"
+#include "log_record.hpp"
+#include "system_parameter.h"
+#include "tde.h"
+
+class corruption_checker
+{
+    //////////////////////////////////////////////////////////////////////////
+    // Does log data sanity checks and saves the results.
+    //////////////////////////////////////////////////////////////////////////
+
+  public:
+    static constexpr size_t IO_BLOCK_SIZE = 4 * ONE_K;   // 4k
+
+    corruption_checker ();
+
+    // Check if given page checksum is correct. Otherwise page is corrupted
+    void check_page_checksum (const LOG_PAGE *log_pgptr);
+    // Additional checks on log record:
+    void check_log_record (const log_lsa &record_lsa, const log_rec_header &record_header, const LOG_PAGE *log_page_p);
+    // Detect the address of first block having only 0xFF bytes
+    void find_first_corrupted_block (const LOG_PAGE *log_pgptr);
+
+    bool is_page_corrupted () const;		      // is last checked page corrupted?
+    const log_lsa &get_first_corrupted_lsa () const;  // get the address of first block full of 0xFF
+
+  private:
+    const char *get_block_ptr (const LOG_PAGE *page, size_t block_index) const;	  // the starting pointer of block
+
+    bool m_is_page_corrupted = false;		      // true if last checked page is corrupted, false otherwise
+    LOG_LSA m_first_corrupted_rec_lsa = NULL_LSA;     // set by last find_first_corrupted_block call
+    std::unique_ptr<char[]> m_null_block;	      // 4k of 0xFF for null block check
+    size_t m_blocks_in_page_count = 0;		      // number of blocks in a log page
+};
+
+corruption_checker::corruption_checker ()
+{
+  m_blocks_in_page_count = LOG_PAGESIZE / IO_BLOCK_SIZE;
+  m_null_block.reset (new char[IO_BLOCK_SIZE]);
+  std::memset (m_null_block.get (), LOG_PAGE_INIT_VALUE, IO_BLOCK_SIZE);
+}
+
+void
+corruption_checker::check_page_checksum (const LOG_PAGE *log_pgptr)
+{
+  m_is_page_corrupted = !logpb_page_has_valid_checksum (log_pgptr);
+}
+
+void corruption_checker::find_first_corrupted_block (const LOG_PAGE *log_pgptr)
+{
+  m_first_corrupted_rec_lsa = NULL_LSA;
+  for (size_t block_index = 0; block_index < m_blocks_in_page_count; ++block_index)
+    {
+      if (std::memcmp (get_block_ptr (log_pgptr, block_index), m_null_block.get (), IO_MAX_PAGE_SIZE) == 0)
```

Comment:
```suggestion
      if (std::memcmp (get_block_ptr (log_pgptr, block_index), m_null_block.get (), IO_BLOCK_SIZE) == 0)
```
Oh, this could be refactored in this issue.

Reply:
done; I hope I got it right